src/otk/transform.py

```diff
@@ -112,15 +112,45 @@ def is_directive(needle: Any) -> bool:
     return isinstance(needle, str) and needle.startswith(PREFIX)
 
 
-@tree.must_be(dict)
-def define(ctx: Context, tree: Any) -> Any:
-    """Takes an `otk.define` block (which must be a dictionary and registers
-    everything in it as variables in the context."""
-
-    for key, value in tree.items():
-        ctx.define(key, value)
+def process_defines(ctx: Context, state: State, tree: Any):
+    """
+    Processes tree for new defines, resolving any references to other variables,
+    and update the global context. The State holds a reference to the nested
+    defines block the function is working in. New defines are added to the
+    nested block but references are resolved from the global ctx.defines.
+    """
 
-    return tree
+    subblock = state.defines
+
+    # Iterate over a copy of the tree so that we can modify it in-place.
+    for key, value in tree.copy().items():
```

Comment:
The commit message for this new recursive function has some very nice examples - I wonder if those could be (easily?) added as test-cases here? Having this commit without tests makes me slightly uneasy (I know tests are coming later/are part of the pre-req but I wonder it it still might be useful to illustrate the new code?)

[edit: tests are in a separate commit, I guess that is fine and just personal preference (to me they are kinda the same workstream so having them in the same commit seems fine but I don't mind to much either way)]

Reply:
Squashed the two commits.